hub/core/chunk_engine/generator.py

```diff
@@ -0,0 +1,95 @@
+import numpy as np
+from typing import Generator, Tuple
+
+from hub.core.chunk_engine.exceptions import ChunkGeneratorError
+
+
+def chunk(
+    content_bytes: bytes, previous_num_bytes: int, chunk_size: int
+) -> Generator[Tuple[bytes, int], None, None]:
+    """
+    Generator function that chunks bytes.
+
+    Chunking is the process of taking the input `content_bytes` & breaking it up into smaller chunks,
+    the sizes of which are <= `chunk_size`.
+
+    Example 1:
+        content_bytes = b"101010101010"
+        chunk_size = 4
+        output_chunks = [b"1010", b"1010", b"1010"]
+
+        This is considered a "perfect fit" because all bytes fit perfectly into 3 chunks.
+
+
+    Example 2:
+        content_bytes = b"10101010111122"
+        chunk_size = 4
+        output_chunks = [b"1010", b"1010", b"1111", b"22"]
+
+        As you can see, the last chunk in `output_chunks` only has 2 bytes (chunk_size=4), which means
+        this is not a perfect fit, but rather it has a partial chunk at the end.
+
+
+    Args:
+        content_bytes(bytes): Bytes object with the data to be chunked.
+        previous_num_bytes(int): The number of bytes in the previous chunk. This is helpful for filling the previous chunk before starting a new one.
+        chunk_size(int): Each individual chunk will be assigned this many bytes maximum.
+
+    Yields:
+        Each yield is a chunk of the `content_bytes`. Each chunk is of length (0, `chunk_size`].
+
+    Raises:
+        ChunkGeneratorError: If leftover bytes are negative or the previous chunk was invalid.
+    """
+
+    if len(content_bytes) <= 0:
+        return
+
+    if previous_num_bytes is None:
+        bytes_left = 0
+    else:
+        bytes_left = chunk_size - previous_num_bytes
+
+    content_num_bytes = len(content_bytes)
+
+    if bytes_left < 0:
+        raise ChunkGeneratorError(
+            "Previous chunk exceeded `chunk_size` (%i > %i)." % (bytes_left, chunk_size)
+        )
+
+    # handle filling the rest of the previous chunk
+    total_bytes_yielded = 0
+    if bytes_left > 0:
+        content_bytes_piece = content_bytes[:bytes_left]
+        yield content_bytes_piece, 0
+        total_bytes_yielded += bytes_left
+
+        if bytes_left > content_num_bytes:
+            # if the previous chunk still isn't filled, that means there is no more
+            # data to write
+            return
+
+    num_chunks_to_create = max(1, int(np.floor(content_num_bytes / chunk_size)))
```

Comment:
`num_chunks_to_create = max(1, content_num_bytes // chunk_size)`  would be easier to read
Btw I think that-
`num_chunks_to_create = math.ceil(content_num_bytes / chunk_size)` makes more sense.
For eg. if there are 8 bytes and chunk_size is 3. We should create ceil(8/3) chunks i.e 3 chunks (3 + 3 + 2)

Reply:
i had `num_chunks_to_create = max(1, content_num_bytes // chunk_size)` originally, but we need to take the floor of this division because if there are leftover bytes, this is not a new full chunk, but a partial chunk. 

i think the solution here is to rename my variable `num_chunks_to_create` to `num_full_chunks`.

Reply:
given your `end = min((piece_index + 1) * chunk_size + bytes_left, len(content_bytes))` solution, im not quite sure if this will work but i will try it!

Reply:
isn't np.floor(a/b) always equal to a//b?

Reply:
ohh i thought it was round division! nice

Reply:
ceil does make more sense!